presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java

```diff
@@ -0,0 +1,102 @@
+import java.util.Map;
+
+import static com.facebook.presto.metadata.Signature.comparableTypeParameter;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.type.ArrayType.toStackRepresentation;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static com.facebook.presto.type.TypeUtils.parameterizedTypeName;
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+
+public final class ArrayIntersectFunction
+        extends ParametricScalar
+{
+    public static final ArrayIntersectFunction ARRAY_INTERSECT_FUNCTION = new ArrayIntersectFunction();
+    private static final String FUNCTION_NAME = "array_intersect";
+    private static final Signature SIGNATURE = new Signature(FUNCTION_NAME, ImmutableList.of(comparableTypeParameter("E")), "array<E>", ImmutableList.of("array<E>", "array<E>"), false, false);
+    private static final MethodHandle METHOD_HANDLE = methodHandle(ArrayIntersectFunction.class, "intersect", Slice.class, Slice.class);
+    public static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider().get().registerModule(new SimpleModule().addSerializer(Slice.class, new SliceSerializer()));
+    public static final CollectionType COLLECTION_TYPE = OBJECT_MAPPER.getTypeFactory().constructCollectionType(List.class, Object.class);
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Intersection of two given arrays.";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        checkArgument(types.size() == 1, format("%s expects only two arguments", FUNCTION_NAME));
+        TypeSignature typeSignature = parameterizedTypeName("array", types.get("E").getTypeSignature());
+        Signature signature = new Signature(FUNCTION_NAME, typeSignature, typeSignature, typeSignature); //return type and arg types are the same
+        return new FunctionInfo(signature, getDescription(), isHidden(), METHOD_HANDLE, isDeterministic(), false, ImmutableList.of(false, false));
+    }
+
+    public static Slice intersect(Slice leftArray, Slice rightArray)
+    {
+        List<Object> leftElements;
+        List<Object> rightElements;
+        try {
+            leftElements = OBJECT_MAPPER.readValue(leftArray.getInput(), COLLECTION_TYPE);
+            rightElements = OBJECT_MAPPER.readValue(rightArray.getInput(), COLLECTION_TYPE);
+        }
+        catch (IOException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, e);
+        }
+        leftElements.retainAll(rightElements);
```

Comment:
Also, this is going to be O(n^2) which is too slow. You should optimize it to use a hash map, so that it runs in linear time